### ISSUE TITLE
Rule Change: no-constant-condition should also block eqeqeq with new object instance

### ISSUE THREAD

**Issue by hfhchan-plb** [eslint/eslint#16923]
### What rule do you want to change?

no-constant-condition

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
if (foo.bar === new Number(baz)) {
 // should trigger: Unexpected constant condition.
}
if (foo.bar === new Boolean(baz)) {
 // should trigger: Unexpected constant condition.
}
if (foo.bar === new String(baz)) {
 // should trigger: Unexpected constant condition.
}
if (foo.bar === new Date(baz)) {
 // should trigger: Unexpected constant condition.
}
if (foo.bar === new Set(baz)) {
 // should trigger: Unexpected constant condition.
}
if (foo.bar === new Map(baz)) {
 // should trigger: Unexpected constant condition.
}
```

### What does the rule currently do for this code?

It can't detect that this always evaluates to false.

### What will the rule do after it's changed?

It will warn / error all eqeqeq to new objects as false. 

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by Rec0iL99** [comment]
Hi @hfhchan-plb, thanks for opening the issue. I'm not able to clearly understand the issue.

From my limited knowledge, I think ESLint will not throw an error in the code example because it does not know the value.

A small example:

This always evaluates to false and ESLint will throw the no-constant-condition because 1 and "1" are literals which are clear to ESLint.

```js
if (1 === "1") {

}
```

Let's take the same example, but put the literals into variables. Now, ESLint will not throw `no-constant-condition` because it does not know the value of `foo` and `bar`. The values `foo` and `bar`could change over time thus making the condition evaluate to `true` or `false`.

```js
const foo = 1
const bar = "1"

if (foo === bar) {

}
```

Here's a [playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWNvbnN0YW50LWNvbmRpdGlvbjogXCJlcnJvclwiKi9cblxuaWYgKDEgPT09IFwiMVwiKSB7XG4gIC8vIHRocm93cyB0aGUgbm8tY29uc3RhbnQtY29uZGl0aW9uXG59XG5cbmNvbnN0IGZvbyA9IDFcbmNvbnN0IGJhciA9IFwiMVwiXG5cbmlmIChmb28gPT09IGJhcikge1xuICAvLyB3b24ndCB0aHJvdyB0aGUgbm8tY29uc3RhbnQtY29uZGl0aW9uXG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczYiOnRydWV9fX0=) demonstrating this.

**Comment by mdjermanovic** [comment]
[`no-constant-binary-expression`](https://eslint.org/docs/latest/rules/no-constant-binary-expression) rule warns about comparisons to newly constructed objects, here's a [playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiaWYgKGZvby5iYXIgPT09IG5ldyBOdW1iZXIoYmF6KSkge1xuIFxufVxuaWYgKGZvby5iYXIgPT09IG5ldyBCb29sZWFuKGJheikpIHtcbiBcbn1cbmlmIChmb28uYmFyID09PSBuZXcgU3RyaW5nKGJheikpIHtcbiBcbn1cbmlmIChmb28uYmFyID09PSBuZXcgRGF0ZShiYXopKSB7XG4gXG59XG5pZiAoZm9vLmJhciA9PT0gbmV3IFNldChiYXopKSB7XG4gXG59XG5pZiAoZm9vLmJhciA9PT0gbmV3IE1hcChiYXopKSB7XG4gXG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1jb25zdGFudC1iaW5hcnktZXhwcmVzc2lvbiI6WyJlcnJvciJdfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==) with the examples from the original post. Since these errors are reported by another rule, there's no need to update `no-constant-condition`.

**Comment by hfhchan-plb** [comment]
Oh right. I'm very sorry that I missed this.
